pager/src/main/java/dev/chrisbanes/accompanist/pager/PagerState.kt

```diff
@@ -0,0 +1,318 @@
+import androidx.compose.runtime.setValue
+import kotlin.math.absoluteValue
+import kotlin.math.floor
+import kotlin.math.roundToInt
+
+private const val LogTag = "PagerState"
+
+/**
+ * Creates a [PagerState] that is remembered across compositions.
+ *
+ * Changes to the provided initial values will **not** result in the state being recreated or
+ * changed in any way if it has already been created.
+ *
+ * @param currentPage the initial value for [PagerState.currentPage]
+ * @param currentPageOffset the initial value for [PagerState.currentPageOffset]
+ * @param pageCount the initial value for [PagerState.pageCount]
+ */
+@ExperimentalPagerApi
+@Composable
+fun rememberPagerState(
+    currentPage: Int = 0,
+    @FloatRange(from = 0.0, to = 1.0) currentPageOffset: Float = 0f,
+    pageCount: Int = 0,
+): PagerState = rememberSaveable(saver = PagerState.Saver) {
+    PagerState(
+        currentPage = currentPage,
+        currentPageOffset = currentPageOffset,
+        pageCount = pageCount
+    )
+}
+
+/**
+ * A state object that can be hoisted to control and observe scrolling for [Pager].
+ *
+ * In most cases, this will be created via [rememberPagerState].
+ *
+ * @param currentPage the initial value for [PagerState.currentPage]
+ * @param currentPageOffset the initial value for [PagerState.currentPageOffset]
+ * @param pageCount the initial value for [PagerState.pageCount]
+ */
+@ExperimentalPagerApi
+class PagerState(
+    currentPage: Int = 0,
+    @FloatRange(from = 0.0, to = 1.0) currentPageOffset: Float = 0f,
+    pageCount: Int = 0,
+) {
+    private var _pageCount by mutableStateOf(pageCount)
+    private var _currentPage by mutableStateOf(currentPage)
+    private val _currentPageOffset = mutableStateOf(currentPageOffset)
+    internal var pageSize by mutableStateOf(0)
+
+    /**
+     * The index of the currently selected page.
+     */
+    @get:IntRange(from = 0)
+    var pageCount: Int
+        get() = _pageCount
+        set(@IntRange(from = 0) value) {
+            _pageCount = value.coerceAtLeast(0)
+            currentPage = currentPage.coerceIn(0, lastPageIndex)
+        }
+
+    internal val lastPageIndex: Int
+        get() = (pageCount - 1).coerceAtLeast(0)
+
+    /**
+     * The index of the currently selected page.
+     *
+     * To update the scroll position, use [scrollToPage] or [animateScrollToPage].
+     */
+    @get:IntRange(from = 0)
+    var currentPage: Int
+        get() = _currentPage
+        private set(value) {
+            _currentPage = value.coerceIn(0, _pageCount)
+        }
+
+    /**
+     * The current offset from the start of [currentPage], as a fraction of the page width.
+     *
+     * To update the scroll position, use [scrollToPage] or [animateScrollToPage].
+     */
+    @get:FloatRange(from = 0.0, to = 1.0)
+    var currentPageOffset: Float
+        get() = _currentPageOffset.value
+        private set(value) {
+            _currentPageOffset.value = value.coerceIn(
```

Comment:
there are no similar verifications for the initial values passed to constructor

Reply:
Yep this is on purpose, to enable state restoration. Although I need to test this. Added a TODO to the PR